refactor!: break down login modal into separate components

sethidden opened the pull request:
Before this PR, LoginModal was a big component which contained 3 forms in a single template.

After this PR, there are 3 separate components for forms, and one parent component for all 3 forms which handles the api queries etc.

Added tests aswell.

atagisow-alokai committed:
refactor: break down login modal into separate components

and add unit tests

fix: loading state not appearing

refactor: add types for forms

fix: squash this

test: no pinia needed here

refactor: minor renamings

chore: update @testing-library pkgs so tests pass

test: remove old login modal tests

refactor: add registerfragment tests and fix misc issues

test: don't mutate vee-validate rules

this is not ideal, as changing the validation translation label would
require chaing the tests, but this is a job for M2-746, where
vee-validat extends() could be moved to a Nuxt plugin

test: add forgot password test

refactor: remove unused keys from fragments

test: add more tests for forgotpassword

refactor: move loginmodal to customer module

chore: flatten test code whitespace

fix: loginmodal path

fix: broken paths after move to customer module

refactor: add login modal tests

refactor: re-add login state

refactor: remove unnecessary extend()

refactor: use less verbose mapper name

refactor: add translation function

refactor: unify to use named import

refactor: use _ instead of _props

sethidden commented on packages/theme/components/LoginModal/__tests__/LoginFragment.spec.ts:
```diff
@@ -0,0 +1,53 @@
+import { required, email } from 'vee-validate/dist/rules';
+import LoginFragment from '../LoginFragment.vue';
+
+extend('required', { ...required, message: 'error:required' });
```
This is not ideal as it makes the test dependent on implementation

sethidden replied:
gotta browse https://vee-validate.logaretm.com/v4/guide/testing more to find a better solution

sethidden replied:
removed the extensions and relying on hardcoded strings for now

sethidden commented on packages/theme/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,244 @@
+<template>
```
remember to move all of those to the customer module

sethidden replied:
done

sethidden commented on packages/theme/components/LoginModal/RegisterFragment.vue:
```diff
@@ -0,0 +1,168 @@
+          :rules="{ required: { allowFalse: false } }"
+        >
+          <SfCheckbox
+            v-model="formCopy.isWillToCreateAccountConfirmed"
```
This shouldn't be part of the emitted payload

sethidden replied:
I'll keep it for now and remove it in the parent LoginModal

sethidden commented:
Need to check if all the :loading props and the error slots work fine

sethidden commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+    };
+
+    const onLoginFormSubmit = async (form: LoginFormFields) => {
+      loginForm.value = form;
+      const { token, cleanup } = await getRecaptchaInfo(isRecaptchaEnabled.value);
+      await login({ user: { ...form, ...(token ? { recaptchaToken: token } : {}) } });
+      cleanup();
+      if (!userError.value.login) {
+        emit('close');
+        await Promise.all([loadItemsCount(), loadCart()]);
+      }
+    };
+
+    const onRegisterFormSubmit = async (form: RegisterFormFields) => {
+      registerForm.value = form;
+      const { token, cleanup } = await getRecaptchaInfo(isRecaptchaEnabled.value);
+      await register({
+        user: {
+          email: form.email,
+          firstname: form.firstName,
+          lastname: form.lastName,
+          password: form.password,
+          is_subscribed: form.shouldSubscribeToNewsletter,
+          ...(token ? { recaptchaToken: token } : {}),
+        },
+      });
+      cleanup();
+      if (!userError.value.register) {
+        emit('close');
+        await Promise.all([loadItemsCount(), loadCart()]);
+      }
+    };
+
+    const onForgotPasswordFormSubmit = async (form: ForgotPasswordFormFields) => {
+      forgotPasswordForm.value = form;
+      const { token, cleanup } = await getRecaptchaInfo(isRecaptchaEnabled.value);
+      await resetPassword({ email: form.username, ...(token ? { recaptchaToken: token } : {}) });
+      cleanup();
+      currentlyDisplayedForm.value = 'forgotPasswordThankYou';
+    };
```
There is a bit of duplication here but I prefer duplication to making some weird abstraction/function.

There's an abstraction now in LoginModal on develop and it just doesn't work - it's used for login and register, but a separate function exists for forgotpassword as the abstraction doesn't fit.

sethidden commented on packages/theme/modules/customer/components/LoginModal/__tests__/LoginModal.spec.ts:
```diff
@@ -0,0 +1,171 @@
+      password: 'P@ssw0rd!',
+    };
+    await waitFor(() => {
+      (loginForm as HTMLElementWithVue).__vue__.$emit('submit', form);
```
Note I'm accessing the Vue instance which is authors of testing-library/vue don't intend me to do. The reason I'm doing it is that I've already tested the form behavior in LoginForm.spec.ts and I know it emits submit correctly.

Basically, I wanted it to be more of a unit test rather than an integration test.

Frodigo replied:
It's not about being more of a unit test rather than an integration test. It's about showing/ not showing implementation details in tests.

sethidden commented on packages/theme/modules/customer/components/LoginModal/forms/__tests__/ForgotPasswordForm.spec.ts:
```diff
@@ -0,0 +1,29 @@
+import ForgotPasswordFragment from '../ForgotPasswordForm.vue';
+import { ForgotPasswordFormFields } from '../types';
+
+describe('ForgotPasswordForm', () => {
```
It's easy to test forms thanks to the props approach.

sethidden commented on packages/theme/modules/customer/components/LoginModal/forms/__tests__/RegisterForm.spec.ts:
```diff
@@ -0,0 +1,22 @@
+import { RegisterFormFields } from '../types';
+
+describe('RegisterForm', () => {
+  it('emits submit if validation passes', async () => {
```
TODO: add more register form tests

bartoszherba commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+import ForgotPasswordThankYou from './forms/ForgotPasswordThankYou.vue';
+import { ForgotPasswordFormFields, LoginFormFields, RegisterFormFields } from './forms/types';
+
+extend('email', {
```
Do we need these extensions here? We have them in forms, and in the LoginModal there is no form nor validator.

sethidden replied:
Right. Removed in f902058080812c5dd614f223cab5193a4474be95

bartoszherba commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+import { extend } from 'vee-validate';
+import { required, email } from 'vee-validate/dist/rules';
+import { useUiState } from '~/composables/useUiState';
+import useCart from '~/modules/checkout/composables/useCart';
```
Please, once you use  { } once you do not. Unify this. (compare useCart and useUser import)

bartoszherba commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+    ForgotPasswordForm,
+    ForgotPasswordThankYou,
+  },
+  setup(_props, { emit }) {
```
why props are underscored? if you are not using props use just `_`
```suggestion
  setup(_, { emit }) {
```

sethidden replied:
I don't think this is important as there's no difference, but I changed it in c3e6870daaea409e8d30683a721b59feaf8e103f because it's quicker than arguing

bartoszherba replied:
Probably, but this is a common approach in many docs, blogs, etc.

bartoszherba commented on packages/theme/modules/customer/components/LoginModal/forms/ForgotPasswordThankYou.vue:
```diff
@@ -0,0 +1,25 @@
+        <slot name="email" />
+      </span>
+    </i18n>
+    <p>{{ $t('Thank You Inbox') }}</p>
```
What is this? ;_) What does it mean?

sethidden replied:
I moved it over from the old LoginModal, I did not change the labels, unless by mistake

sethidden replied:
I think the original author meant "inbox message (thanks, check your inbox) for the forgot password thank you page"

bartoszherba commented on packages/theme/modules/customer/components/LoginModal/forms/ForgotPasswordThankYou.vue:
```diff
@@ -0,0 +1,25 @@
+  </div>
+</template>
+
+<script lang="ts">
```
Is this required?
```
<script lang="ts">
import { defineComponent } from '@nuxtjs/composition-api';

export default defineComponent({});
</script>
```

bartoszherba replied:
No, it is not. Neither in vscode or webstorm. But what is highlighted in both is (regardless of the defineComponent).
![Uploading Screenshot 2022-06-08 at 08.06.24.png…]()

sethidden replied:
<img width="705" alt="image" src="https://user-images.githubusercontent.com/5359825/172553426-958f5fa9-48e1-410a-be80-9cbc99b3bce5.png">

whereas it's fine with defineComponent

<img width="518" alt="image" src="https://user-images.githubusercontent.com/5359825/172553608-d754e9e2-1824-4920-b4c0-ac79bd1af9ce.png">

I assume you're using Vetur, not Volar? (webstorm has its own inhouse extension).

Having the script isn't a big deal. I'll resolve this.

bartoszherba commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+          data-testid="register-form"
+          :show-recaptcha="isRecaptchaEnabled"
+          :loading="loading"
+          :form="registerForm"
```
Ok, it is easier to test form but I still do not understand what do we have formCopy and watcher.

sethidden replied:
In a situatioin where fromCopy doesn't exist, I'd be forced to something like this:

```
props: { form: { type: Object as PropType<...> }

<input v-model="form.lastName"
```

this obviously equivalent to `<input :value="form.lastName" @input="form.lastName = $event`

So as you can see, this would be modifying props directly, which is not only illegal, but throws erros in the console

This is also necessary because I need some variable to buffer the changes made by the user
1. inital form is passed to LoginForm
2. user makes some changes in the form, but those changes exist only in formCopy
3. once the form validation lets me know that what the user typed in is fine, only then I emit formCopy, which replaces the real, non-copy loginForm ref

sethidden commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+          :loading="loading"
+          :form="loginForm"
+          @submit="form => onLoginFormSubmit(form)"
+          @go-to-register="currentlyDisplayedForm = 'register'"
```
I thought about this while writing this code. But I decided that I'll keep it as-is because I don't want the Login**Form** to know anything about Login**Modal** logic (currently displayed form). LoginForm is just a form - has some inputs then emits an object - there's no reason for it to hold "currentlyDisplayedForm" logic. Clicking the "go to register" button emits the go to register event

What the parent component does with that event is up to it and the logic it handles

As for extracting to function, there's too little logic here to warrant extracting. I prefer the at-a-glance assignment in the template since you don't have to look at the method to see what it does.

I've thout about maybe changing go-to-register to navigate:register or something

bartoszherba replied:
Now  LoginForm knows about the registration form and even more, it expects the register form and even makes sense if and only when RegisterForm exists. On the other hand, a generic go-to event assumes nothing about anything and can be eventually handled by ANY parent in ANY way. go-to-register is more like a well-defined command than an event and go-to event emitter knows nothing about any other component as it can be handled anywhere not just in LoginModal

Frodigo replied:
I agree with Bartosz here, once we use events like `@go-to-register`, `@go-to-forgot-password`, etc, we need to add a new event when we want to add a new form (or component) inside the component. When we use the `@go-to` event with an argument, it makes the LoginModal component more generic and readable.

sethidden replied:
fa43904764991d4d5204dde500fb0fbc11ae80c5

sethidden commented on packages/theme/modules/customer/components/LoginModal/LoginModal.vue:
```diff
@@ -0,0 +1,251 @@
+    >
+      <div>
+        <LoginForm
+          v-if="currentlyDisplayedForm === 'login'"
```
As in previous comment

>As for extracting to function, there's too little logic here to warrant extracting. I prefer the at-a-glance assignment in the template since you don't have to look at the method to see what it does.

bartoszherba replied:
Than is why functions can be named ;)

atagisow-alokai committed:
refactor: use one navigation event instead of many

I don't like this commit because this couples the loginmodal logic to
the respective forms (they shouldn't know anything about loginmodal's
page logic), but reviewers think otherwise so
